### ISSUE TITLE
Calc with viewport units seems to be broken.

### ISSUE THREAD

**Issue by emilio** [servo/servo#12814]
Found in #12813. 

``` html
<!doctype html>
<style>
  div {
    width: 100%;
    height: calc(100vh - 10px);
    background: red;
  }
  html, body { margin: 0; padding: 0; }
</style>
<div></div>
```

Expected: All screen red except the ten rows of pixels of the bottom.
Actual result: All white.

**Comment by emilio** [comment]
Putting the tag on CSS because it's likely to be it, I have no time right now to dig more, but I doubt a lot it's layout related.

**Comment by jdm** [comment]
I wonder if this is caused by treating the viewport as the size of the body content?

**Comment by emilio** [comment]
Hmm... So `div { height: 100vh; }` does paint everything red, yet `div { height: calc(100vh); }` doesn't. And it's even more funny, and probably related to what @jdm is proposing, that any of the calc examples work iff you give the `body` element an explicit height, even if it's a `1px` one.

**Comment by emilio** [comment]
Fixed by #16541
